src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,243 @@
+        category_assoc_types.empty? || category_assoc_types.include?("VirtualMachine")
+      end
+
+      def create_tag_hash(vm_config_tags)
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+          begin
+            raise CreateTagHashCategoryError.new("Missing category content") if vm_config_tag["category"].nil?
+            raise CreateTagHashCategoryError.new("Empty category") if vm_config_tag["category"].empty?
+            raise CreateTagHashTagError.new("Missing tag",vm_config_tag["category"] ) if vm_config_tag["tag"].nil?
+            raise CreateTagHashTagError.new("Empty tag",vm_config_tag["category"] ) if vm_config_tag["tag"].empty?
+          rescue => e
+            if e.instance_of?(CreateTagHashCategoryError)
+              logger.warn("Create Tag Hash Category Error Raised with message : #{e.message}")
+            else
+              logger.warn("Create Tag Hash Tag Error Error Raised with message : #{e.message}")
+            end
+            next
+          end
+          tag_hash[vm_config_tag["category"]] ||= []
+          tag_hash[vm_config_tag["category"]] << vm_config_tag["tag"]
+        end
+        tag_hash
+      end
+
+      def attach_single_tag(vm_mob_id, tag_id)
+        tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttach.new
+        tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        tag_assoc_info.object_id.id = vm_mob_id
+        tag_assoc_info.object_id.type = 'VirtualMachine'
+        tag_association_api.attach(tag_id, tag_assoc_info)
+      end
+
+      def attach_multi_tags(vm_mob_id, tag_ids, category_name, target_category_id)
+        if tagging_category_api.get(target_category_id).value.cardinality == "SINGLE"
+          logger.warn(CardinalityError.new(category_name))
+          return
+        end
+        multi_tag_assoc_info = VSphereAutomation::CIS::CisTaggingTagAssociationAttachMultipleTagsToObject.new
+        multi_tag_assoc_info.object_id = VSphereAutomation::CIS::VapiStdDynamicID.new
+        multi_tag_assoc_info.object_id.id = vm_mob_id
+        multi_tag_assoc_info.object_id.type = 'VirtualMachine'
+        multi_tag_assoc_info.tag_ids = tag_ids
+        tag_association_api.attach_multiple_tags_to_object(multi_tag_assoc_info)
+      end
+
+      def attach_tags(vm_mob_id, vm_config_tags, vm_config_name)
+        tag_hash = create_tag_hash(vm_config_tags)
+        if tag_hash.empty?
+          logger.warn(NoCategoryTagPairError.new)
+          return
+        end
+
+        category_ids = tagging_category_api.list.value
+        if category_ids.empty?
+          logger.warn(VCenterNoCategoryExitsError.new)
+          return
+        end
+        tag_hash.keys.each do |category_name|
+          begin
+            target_category_id = retrieve_category_id(category_name, category_ids)
+            raise VCenterBadEntityError.new(category_name, "category") if target_category_id.nil?
```

Comment:
Why is it not `VCenterNoCategoryExitsError`?

Reply:
Please refer to explain for line 8.

Here is a copy : 

VCenterNoCategoryExitsError avoid to traverse the categories in tag_hash, when there is no categories on vCenter, we just finish attach tag steps.

Similar with VCenterNoTagFoundError, when try to attach one or more tags in a category, we first use the category id got a list of tags under it, when the list is empty, throw out the message and skip check tags in category.

VCenterBadEntityErrorrefers to a single category or tag check result.

These 3 throws different messages, that why we need 3 classes.